### PR TITLE
Annotate platform specific hardware intrinsics with the ConstExpected attribute

### PR DESCRIPTION
**Pull request by tannergooding**
The xplat intrinsics do not need this as they use alternative instruction sequences and handling where possible. They do not need to fallback to anything like a jump table.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/AdvSimd.cs [hunk 1]
```diff
@@ -51,25 +51,25 @@ public abstract class AdvSimd : ArmBase
         //     /// float32x2_t vmla_lane_f32 (float32x2_t a, float32x2_t b, float32x2_t v, const int lane)
         //     ///   A32: VMLA.F32 Dd, Dn, Dm[lane]
         //     /// </summary>
-        //     public static Vector64<float> MultiplyAddBySelectedScalar(Vector64<float> addend, Vector64<float> left, Vector64<float> right, byte rightIndex) => MultiplyAddBySelectedScalar(addend, left, right, rightIndex);
+        //     public static Vector64<float> MultiplyAddBySelectedScalar(Vector64<float> addend, Vector64<float> left, Vector64<float> right, [ConstantExpected(Max = (byte)(1))] byte rightIndex) => MultiplyAddBySelectedScalar(addend, left, right, rightIndex);
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/AdvSimd.cs:54]: Arm64 and WASM expect values to be exactly in range. As such, we strictly limit the input between `0` and `Max` or `1` and `Max`.

The exact semantics is dependent on the instruction and is table driven in the JIT: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/hwintrinsicarm64.cpp#L215-L271

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/X86/Avx2.cs [hunk 2]
```diff
@@ -802,56 +803,56 @@ public new abstract class X64 : Avx.X64
...
-        public static new Vector128<ulong> ExtractVector128(Vector256<ulong> value, byte index) => ExtractVector128(value, index);
+        public static new Vector128<ulong> ExtractVector128(Vector256<ulong> value, [ConstantExpected] byte index) => ExtractVector128(value, index);
 
         /// <summary>
         /// __m128i _mm_i32gather_epi32 (int const* base_addr, __m128i vindex, const int scale)
         ///   VPGATHERDD xmm, vm32x, xmm
         /// The scale parameter should be 1, 2, 4 or 8, otherwise, ArgumentOutOfRangeException will be thrown.
         /// </summary>
-        public static unsafe Vector128<int> GatherVector128(int* baseAddress, Vector128<int> index, byte scale)
+        public static unsafe Vector128<int> GatherVector128(int* baseAddress, Vector128<int> index, [ConstantExpected(Min = (byte)(1), Max = (byte)(8))] byte scale)
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/X86/Avx2.cs:855]: The one particularly interesting case for x86/x64 is `GatherVector` and `GatherMaskVector`. These take exactly `1`, `2`, `4`, or `8` with other values being disallowed.

This is most closely modeled on the attribute by restricting to between `1` and `8` inclusive.

**Review by dakersnar — COMMENTED** [review]
Were these annotations all manually added? What were your source of truth for the expected constants in each annotation? How (if at all) are these tested?

**Comment by tannergooding** [comment]
> Were these annotations all manually added?

I wrote a small program that added them.

> What were your source of truth for the expected constants in each annotation?

The existing JIT conditions, which were originally seeded from the Architecture Manuals

> How (if at all) are these tested?

Manual validation in VS that the analyzer gets triggered

**Review by dakersnar — APPROVED** [review]
Thanks for the clarifications, this LGTM at a high level.

**Comment by tannergooding** [comment]
The fix is easy, but I'm unsure why it didn't trigger these failures in the PR.

**Comment by tannergooding** [comment]
Fix is here: https://github.com/dotnet/runtime/pull/80282

**Pull request merged**